File: src/org/genyris/java/swing/ListenerFunction.java
package org.genyris.java.swing;


import org.genyris.core.Constants;
import org.genyris.core.Exp;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.java.JavaWrapper;

public class ListenerFunction extends ApplicableFunction {

	public ListenerFunction(Interpreter interp) {
		super(interp, Constants.PREFIX_JAVA + "actionListener", true);
	}

	public Exp bindAndExecute(Closure proc, Exp[] arguments,
			Environment env) throws GenyrisException {
		Class[] types = {Closure.class};
		this.checkArgumentTypes(types, arguments);
		return new JavaWrapper( new GenyrisActionListener((Closure)arguments[0], env));		
	}

}

File: src/org/genyris/java/swing/ListenerFunction.kt
package org.genyris.java.swing

import org.genyris.core.Constants
import org.genyris.core.Exp
import org.genyris.core.PrefixSymbol
import org.genyris.exception.GenyrisException
import org.genyris.interp.ApplicableFunction
import org.genyris.interp.Closure
import org.genyris.interp.Environment
import org.genyris.interp.Interpreter
import org.genyris.java.JavaWrapper


class ListenerFunction(interp: Interpreter) :
    ApplicableFunction(interp, PrefixSymbol(Constants.PREFIX_JAVA, "actionListener", "sys"), true) {
    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(
        proc: Closure?, arguments: Array<Exp?>,
        env: Environment?
    ): Exp {
        val types = arrayOf<Class<*>?>(Closure::class.java)
        this.checkArgumentTypes(types, arguments)
        return JavaWrapper(GenyrisActionListener(arguments[0] as Closure?, env))
    }
}
